initial docs for splitting miner and markets service (#890)
filecoin-project/filecoin-docs: docs/mine/lotus/split-markets-miners.md

---
title: 'Lotus Miner: Splitting main miner and markets service processes'
description: ''
breadcrumb: 'Splitting main miner and markets service processes'
---

# {{ $frontmatter.title }}

{{ $frontmatter.description }}

## Background

Lotus v1.11.0 introduced the notion of `subsystems` in the `lotus-miner` process. Currently there are 4 subsystems, that can be enabled or disabled in the `config.toml` file:

```
[Subsystems]
#  EnableMining = true
#  EnableSealing = true
#  EnableSectorStorage = true
#  EnableStorageMarket = true
```

All these options are set to `true` by default. Until now, the monolith `lotus-miner` process has been responsible for all functionality.

> At the moment these `subsystems` are designed to be grouped into two distinct types of `lotus-miner` nodes:

Reviewer: What happens if the user passes an unsupported configuration? Could we explain it in the doc?

When a `lotus-miner` node is started with configuration for a `mining/sealing/proving` node, it does not include a libp2p node, and it doesn't have to be publicly exposed on the Internet. The `markets` node communicates with it via its JSON RPC interface.

This document explains how to split your existing monolith `lotus-miner` node into multiple processes and take advantage of this new architecture, which brings more security to your mining operation as well as less potential attack surface.

## Pre-requisites

Before splitting the markets service process from the monolith miner process, you should backup your miner's metadata repository. You need to start the `lotus-miner` with the `LOTUS_BACKUP_BASE_PATH` env variable in order to do that.

```sh
export LOTUS_BACKUP_BASE_PATH=~/lotus-backup-location
lotus daemon
```

```sh
export LOTUS_BACKUP_BASE_PATH=~/lotus-backup-location
lotus-miner run
```

## Creating a backup

```sh
export LOTUS_BACKUP_BASE_PATH=~/lotus-backup-location
lotus-miner backup ~/lotus-backup-location/backupfile
```

## Copy config.toml and storage.json for the markets service

> You need to generate a `config.toml` and `storage.json` for the markets node, and have them ready for the next step.

Reviewer: Link to instructions?

Reviewer: Can we also explain what each file does, as a reminder?

Author: I've linked to the two relevant pages, although I think it is fair to say that miners who attempt to split their monolith, should already be quite familiar with the `config.toml` file.

If you intend to run the `mining/sealing/proving` node on the same machine as the `markets` node, make sure that their listeners do not clash:
```
[API]
  ListenAddress = "/ip4/127.0.0.1/tcp/8787/http"
  RemoteListenAddress = "127.0.0.1:8787"
```

By default the `lotus-miner` node listens to port 2345, so in the example configuration above, we change the configuration for the `markets` node to listen to `127.0.0.1:8787`.

`storage.json` should be left without storage paths, as the `markets` node won't handle anything storage related.

```
{
  "StoragePaths": [
  ]
}

> ```

Reviewer: Why is it necessary then?

Author: Not necessary anymore, and removed.

In the example commands below, we have placed the `config.toml` and `storage.json` in the `~/.lotusmarket` directory.

## Initialising a markets service repository

1. Create authentication tokens for the `markets` node

```sh
export APISEALER=`./lotus-miner auth api-info --perm=admin`
export APISECTORINDEX=`./lotus-miner auth api-info --perm=admin`
```

> 2. Initialise the `market` node

Reviewer: ```suggestion
2. Initialise the `market` node. This performs a one-time setup.
```

```sh

> export LOTUS_MINER_PATH=~/markets-repo-location

Reviewer: Need to make add a warning that the `lotus-miner` command will interact with one or other repo depending on the path!

## Background

> Lotus v1.11.0 introduced the notion of `subsystems` in the `lotus-miner` process. Currently there are 4 subsystems, that can be configured (enabled/disabled) via the `config.toml` file:

Reviewer 2: ```suggestion
Lotus v1.11.0 introduced the notion of `subsystems` in the `lotus-miner` process. Currently there are 4 subsystems, that can be enabled or disabled in the `config.toml` file:
```

```
[Subsystems]
#  EnableMining = true
#  EnableSealing = true
#  EnableSectorStorage = true
#  EnableStorageMarket = true
```

> By default all are set to `true`, since until now the monolith `lotus-miner` process has been responsible for all functionality.

Reviewer 2: ```suggestion
All these options are set to `true` by default. Until now, the monolith `lotus-miner` process has been responsible for all functionality.
```

## Background

> > Note: this feature is to be introduced in Lotus v1.11.1

Reviewer: ```suggestion
> Note: this feature is available starting with Lotus v1.11.1
```

Rationale: the future tense makes sense now, but we won't come and update this when it becomes the past.

export APISECTORINDEX=`./lotus-miner auth api-info --perm=admin`
```

> 2. Initialise the `market` node. This performs a one-time setup, in which the `Peer ID` and the `Multiaddr` of the miner are updated and a message is submitted so that these values are changed on-chain. This is necessary so that storage and retrieval clients know that this miner is now publicly dialable/reachable on a new addres - namely the one that the `market` node is listening to. Note that `lotus-miner` interacts with a given repository depending on the `LOTUS_MINER_PATH` environment variable!

Reviewer: ```suggestion
2. Initialise the `market` node. This performs a one-time setup of the markets node. Part of that setup includes updating the `Peer ID` and the `Multiaddr` in the miner actor by submitting a message on chain. This is necessary so that storage and retrieval clients know that this miner's **deal-making** endpoint is now publicly dialable/reachable on a new address (the new `market` node).

    Note that `lotus-miner` interacts with one repository or another depending on the `LOTUS_MINER_PATH` environment variable!
```

## Start the `mining/sealing/proving` miner process without the markets subsystem

1. Update your `config.toml` and set `EnableStorageMarket` option to `false`.

> 2. Run the process with the `--enable-markets=false` flag

Reviewer 2: ```suggestion
1. Update your `config.toml` and set `EnableStorageMarket` option to `false`.
2. Run the process with the `--enable-markets=false` flag
```
